### ISSUE TITLE
IDE0050 - Convert to tuple should not apply in expression trees

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/roslyn#50468]
Found as part of a query expression that uses an anonymous type. Simplified repro below:

```csharp
Expression<Func<int, int, object>> f = (x, y) => new { x, y };
```

suggests to convert to:

```csharp
Expression<Func<int, int, object>> f = (x, y) => (x, y);
```

which then results in:

```
CS8143	An expression tree may not contain a tuple literal.
```

**Comment by CyrusNajmabadi** [comment]
This one is debatable. Code transforms try to be commonly available so that users can make those changes as part of a larger series of edits. For example, say I realized I did *not* want this to be an expression tree with an anonymous type. I don't want to be forced to change things in a certain order. 

So, in this case, I think offering the transform is fine. Note: this is very different from an analyzer where we are *telling* the user: make this change.  Or where we will go change many locations in a fix-all. 

These transforms are more about doing something quickly that would be annoying to do manually.

**Comment by bartdesmet** [comment]
Then consider the simplest of LINQ 101 queries using an anonymous type as a projection:

```csharp
static void Demo(IQueryable<Product> products)
{
    var res = from p in products
                select new { p.Name, p.Price };

    foreach (var item in res)
    {
    }
}
```

The suggested fix yields invalid code here as well, simply because the `select` clause binds to an expression-tree based `Select` method:

![image](https://user-images.githubusercontent.com/7324746/104638880-982c8800-565b-11eb-8232-e1bf362df712.png)

with the following error:

![image](https://user-images.githubusercontent.com/7324746/104638960-b2666600-565b-11eb-95bd-866631ac2bba.png)

Of course the better thing would be that expression trees get some renewed love from the language designers and don't have to reject anything > C# 3.0, but that's a whole different story :-).

The point of the issue being that a code fix yields invalid code that doesn't compile because of inherent restrictions in the language. I don't think it should even suggest to make such a change.

**Comment by bartdesmet** [comment]
Note there are a whole bunch of these issues with expression trees. Another example is https://github.com/dotnet/roslyn-analyzers/issues/4665. While that one compiles, query providers barf at runtime never having seen `Array.Empty<T>()` where they did understand `new T[0]` before.

In general, it'd be good to consider whether expression trees should be "demilitarized zones" for code fix providers. Given that an API is capturing a lambda expression as an expression tree typically means there's some runtime library performing code analysis, often looking for various patterns (e.g. in query translation). So if you're lucky and the suggested code fix works at compile time (unlike the one shown in this issue, which even fails to compile), you may still end up with a runtime exception due to a change to the captured expression tree.

There were quite a few Roslyn bugs in the early days where query providers broke because seemingly innocent redundant conversions were taken away (just to name [one example](https://github.com/dotnet/roslyn/issues/5734)). Now, we seem to be giving ammunition to users to potentially break their code *at runtime* when whatever runtime library that receives the expression tree (e.g. a query provider) doesn't recognize the new tree structure. (Or worse, the generated code by e.g. a query provider comes to the conclusion it can't execute the query remotely and goes for an alternative execution strategy where part of the query gets executed locally after transferring a bunch more data.)

Another example I just ran into is IDE0004 in expression trees which suggests to take away redundant conversions, which obviously results in `UnaryExpression` nodes of the `Convert` kind to be dropped in the resulting tree. That, too, has the potential of breaking things.

**Comment by bartdesmet** [comment]
Here's another example:

```csharp
Expression<Func<object, bool>> f = o => o is bool ? (bool)o : false;
```

IDE0038 suggests to use pattern matching:

![image](https://user-images.githubusercontent.com/7324746/104641983-47b72980-565f-11eb-8606-a82d322e4f98.png)

The same faulty rewrite would be suggested in "hidden" uses of expression trees, e.g. clauses in query expressions that bind to expression tree based overloads of standard query operators.

**Comment by CyrusNajmabadi** [comment]
> The suggested fix 

These are not 'suggested fixes'.  First, there's nothing to fix there.  Second, we're not actually telling you to do this.  We're just offering shortcuts to speed up transforms that you can *choose* to use if they are what you *want* to do.  Note: this is standard *everywhere* where you type.  You can almost always bring up these shortcuts to help speed things up.  We're not saying: you should do this.  We're saying: if you want to do this as an individual step of work, we're here to make that easier :)

**Comment by CyrusNajmabadi** [comment]
> The point of the issue being that a code fix yields invalid code that doesn't compile because of inherent restrictions in the language. I don't think it should even suggest to make such a change.

Fundamentally, this is not how i think we want to think about these things.  I *commonly* use lots of code-transforms that move my code through various states of invalidity.  I eventually end up in a valid state, but i don't want to be blocked from thse quick transforms just because an intermediary state is *temporarily* invalid.

> 0038 suggests to use pattern matching:

This is a known bug.  Here it's an issue because we're actually *suggesting* to use pattern matching.  We should not *suggest* to make these changes actively with a visual annotation in the text itself.  I believe this is a known issue, but i'm not sure what the bug number is.

**Comment by bartdesmet** [comment]
> I eventually end up in a valid state, but I don't want to be blocked from these quick transforms just because an intermediary state is temporarily invalid.

Well, this is hardly a "temporarily invalid intermediate" state when your code got transformed into little cul-de-sacs (and you're lucky to notice at compile-time; as mentioned above, any rewrite inside expression trees can cause runtime errors) and the only way to get out is to make a U-turn and undo the "change" (because the word "fix" is apparently sensitive in this context). I do agree there are cases where successive transforms are needed to get to a valid state, but in this case I'm dealing with a checkmate.

> The point of the issue being that a code fix yields invalid code that doesn't compile because of inherent restrictions in the language.

Note that I'm talking about *inherent* restrictions to the language. Things I can't do anything about. And, no, I can't change my APIs to not take expression trees to move to my next valid state. Quotation is a strong intent, often forced upon me by query providers and the like.

> We should not suggest to make these changes actively with a visual annotation in the text itself.

Whether it's "suggesting" or not seems a bit mood to me now that we have toggles like:

```
dotnet_analyzer_diagnostic.category-Style.severity = warning
```

and

```
<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
```

which I'm experimenting with right now (knowing that the support for this itself is "experimental") and these effectively become warnings:

```
...\DataSerializerTests.cs(552,28): warning IDE0050: Convert to tuple
```

And surely my IDE correctly reflects my preferences (hooray!) so I can burn down the list of warnings:

![image](https://user-images.githubusercontent.com/7324746/104645277-b4ccbe00-5663-11eb-9586-490b665c95e7.png)

Maybe with these flags you can argue that I'm telling you to now suggest more things, complain about more things, etc. and I'm only to blame myself. I'm surely opting in to another level of pain, but I've evaluated it's worth the time. Except...

For 95% of the time, I *want* the changes, because I want to keep my code base modern and clean. For 4% of the time, I'm happy to *suppress* things where I disagree (either by tweaking a rule's behavior, by disabling a rule altogether, by using `#pragma` in a few places, or some other means to suppress selectively). For the remainder 1% of the time, I'm *frustrated*. In this case, I want to convert to tuples in a bunch of places. It's making my code better, so I don't want to disable the rule. But for cases like the one shown in this issue, I have to suppress the warning, and a `#pragma` is my weapon of choice because I can keep the scope as small as possible and have it travel with the code in case of refactorings. All I have to do is come up with a snarky comment to justify the suppression right there:

![image](https://user-images.githubusercontent.com/7324746/104645530-0117fe00-5664-11eb-8488-3ffa871dc192.png)

It just makes me cringe having to write a justification for something that's just *unfixable because of inherent limitations in the language* that the analyzer seems to have forgotten about. Another example of this was https://github.com/dotnet/roslyn-analyzers/issues/4651 where the suggested-or-not-but-still-shown-to-me transform yielded invalid C# code that merely screams "we don't know what we're doing". Whether it's "hey, make your indexer static" (oh, wait, the language says there's no such thing, so I can't) or "hey, use modern stuff in your expression tree" (oh, wait, expression trees are still living in the C# 3.0 days, so I can't), they're all non-actionable, corner me into a dead end, and I have to throw in a suppression to make it go away.

In short, what's the stance on analyzers offering up code transforms that yield invalid C# (e.g. [static indexers](https://github.com/dotnet/roslyn-analyzers/issues/4651), pattern matching in expression trees, [operator overloads with insufficient parameters](https://github.com/dotnet/roslyn-analyzers/issues/4697), [inserting a cast to `var`](https://github.com/dotnet/roslyn/issues/50469))? Some, like the impossible cast to `var`, are easy to address and small fit-n-finish annoyances. But for others, suppression is the only resort.

**Comment by CyrusNajmabadi** [comment]
> Well, this is hardly a "temporarily invalid intermediate" state

There is no way to know this.  I have made just htese sorts of changes myself, and then made further changes to get to a fully working state.

**Comment by CyrusNajmabadi** [comment]
> It just makes me cringe having to write a justification for something that's just unfixable because of inherent limitations in the language 

I'm really confused.  Why is this something you are even suppressing?  Nothing in the IDE should be encouraging you to do this.  It's just a code transformation feature that should pop up if you explicitly ask: hey, what transforms can you help with here?

**Comment by CyrusNajmabadi** [comment]
> Whether it's "suggesting" or not seems a bit mood to me now that we have toggles like:

@mavasani How can we get this code-transformer to not participate here.  This feature is jsut a quick transform feature (similar to 'invert-if').  It should never cause a warning or anything like that.  Thanks!

**Comment by bartdesmet** [comment]
Follow the instructions at https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview#enable-on-build and you get the whole shebang.

**Comment by CyrusNajmabadi** [comment]
> which I'm experimenting with right now (knowing that the support for this itself is "experimental") and these effectively become warnings:

Ack.  That is hte bug here.  In no way should this feature ever be associated with any sort of severity.  It's just a code transform (like 'invert if').  It's never intended to be something that should force the user to move their anonymous-types to tuples.  Such a thing doesn't even make sense.  Many APIs out there *require* anonymous types.  That would simply break things.

**Comment by CyrusNajmabadi** [comment]
> For 95% of the time, I want the changes, because I want to keep my code base modern and clean

Tuples are not more modern/clean than anonymous-types.  They're just an alternate.  We have the feature to allow you switch back and forth so that if it makes sense to you to use tuples, you can easily do it.  Someone mis-categorized this feature.  That's the bug here :)

**Comment by bartdesmet** [comment]
> Tuples are not more modern/clean than anonymous-types.

Sure, they're alternatives. But for performance-critical code, I want them. Anyway, we digress.

> Someone mis-categorized this feature. That's the bug here :)

Sure, that's one of the bugs. I'm still feeling strong about not offering up changes that produce invalid C#. Not some intermediate state the user can get out of. Fundamentally incorrect C# code. We can also agree to disagree on this; I'm just trying to find out what the position is of the team on what's considered an invalid transform.

**Comment by CyrusNajmabadi** [comment]
> Not some intermediate state the user can get out of. Fundamentally incorrect C# code.

It's not fundamentally incorrect if i'm in the middle of a series of changes that will make things correct.  For example, if i realize i don't want/need expression trees here.  I may also be copy/pasting this from one place to another where i intend to move from expression trees to in-memory.

Note: thi is not a hypothetical.  I've done this *many* times myself.  I've moved whole swaths of expresion-tree code to in-memory code, and i don't want to be forced to do it in a specific order that slows me down.

>  I'm just trying to find out what the position is of the team on what's considered an invalid transform.

It's invalid if we take correct code and make it an error *and* are explicitly suggesting people do this (i.e. with `...` or a warning/error squiggle).

It's not invalid if it's a general transformation speedup.  Those are not expected to be explicitly preferred.  They generally have no associated 'fix all', and they only come up when the user explicitly says: hey, show me the transforms available here.

THen, we try our best to maintain semantics.  But we don't strictly insist that they hold.

**Comment by bartdesmet** [comment]
> It's not fundamentally incorrect if i'm in the middle of a series of changes that will make things correct.

If that's the bar to measure against, a lot of rewrites become valid. "Translate method body to VB." is valid in a C# file then as well. Because I may want to rename the file to .vb in the future.

> I've moved whole swaths of expresion-tree code to in-memory code, and i don't want to be forced to do it in a specific order that slows me down.

So the order in this case is: "oh, I can use tuples (or whatever other feature) here if I get rid of expression trees" (query provider, runtime metaprogramming framework, translator, expression serializer across services, etc.), let's first make the code invalid and then go figure out where to get rid of `Expression<T>` in APIs I may not even own. If the intent for whatever valid reason is to move from an expression tree to "in-memory", then I'd expect that to be the first step because it changes the core intent and execution strategy of the program. And then a whole bunch of "modernizations" (or whatever you want to call them) become valid.

> They generally have no associated 'fix all'

This one has a "fix all" as shown below.

![image](https://user-images.githubusercontent.com/7324746/104654116-7ab5e900-5670-11eb-94e9-4040772b694b.png)

In a big code base (in my case, the "fix all" yielded 397 occurrences, of which a handful were LINQ-related, resulting in this issue report), won't you end up driving people that are using LINQ (e.g. to SQL or EF) and, unlike us, have never heard of expression trees (the whole point of the feature is for users not to have to know about "quotation" going on) into mysterious errors:

```
CS8143	An expression tree may not contain a tuple literal.
```

"What's an expression tree? Isn't this just a LINQ query expression?"

One can hope their next step is to back out the change in that single place rather than doing something silly like turning:

```csharp
var res = from p in products
          select new { p.Name, p.Price };
```

into

```csharp
var res = from p in products.AsEnumerable()
          select (p.Name, p.Price);
```

and downloading the whole database. And that's exactly my point about all sorts of constructs in expression trees getting rewritten, maybe using a "fix all", e.g. `new T[0]` to `Array.Empty<T>()`. One can hope to get a compile-time error to catch these issues; one may get a runtime error and still catch an issue; or one may be unlucky and cause the whole runtime evaluation strategy to get overturned (e.g. silently turn to local evaluation and end up downloading a whole database).

Less than 0.1% of folks may be in the middle of a "let's get rid of expression trees, throw in some other transforms while I'm at it, sure things will have red squiggles" rewrite. The other 99.9% are maintaining an existing code base, hit "fix all" on things, and end up with one of the many `An expression tree may not contain <insert any feature in C# 4.0-9.0 here>.` errors, or have silently altered the shape of a quotation that a library will get faced with later.

To put it bluntly: quotations are *almost* like string literals passed down to a library. Should code transformations be in the business of altering literals?

**Comment by CyrusNajmabadi** [comment]
> If that's the bar to measure against, a lot of rewrites become valid.

Yes.  Agreed.  My position is that we should allow these rewrites.

**Comment by CyrusNajmabadi** [comment]
> So the order in this case is:

Yes.  I routinely do things in whatever order makes sense for me at the time.  I do not want to be forced to try to go in an explicit order when doing a bunch of transforms.

**Comment by CyrusNajmabadi** [comment]
> This one has a "fix all" as shown below.

Yes.  This is a bug.  i'm still waiting on information from another dev on how we can address this.  Under no circumstances should this particular transform proffer UI or behave in this fashion.  That's not at all the intent/design around it.  it appears as if it was unintentionally included into a set of behaviors i twas not intended for.  The fixes all go in that space.

**Comment by CyrusNajmabadi** [comment]
> In a big code base (in my case, the "fix all" yielded 397 occurrences,

There will be no more fix-all for this once i get information on the right way to opt this feature out of that experience.

**Comment by CyrusNajmabadi** [comment]
> The other 99.9% are maintaining an existing code base, hit "fix all" on things, 

Again, 'fix all' is a bug here.  It should not be offered.  There should be nothing explicitly pushing this on the user, or making this feature make lots of changes to the source code.  It's as buggy as if we had an 'fix all - invert if in solution'.  It's not a sensical o intended option for this feature.  You are dealing with the fallout of that brokenness.  The right fix here is to fix the root cause, not attempt to address a symptom caused by the root problem.

**Comment by CyrusNajmabadi** [comment]
> To put it bluntly: quotations are almost like string literals passed down to a library. Should code transformations be in the business of altering literals?

Yes.  Code *transformations* should be in the business of altering those.  Code *fixes* should not be.

**Comment by mavasani** [comment]
@CyrusNajmabadi [IDE0050](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0050) (convert anonymous type to tuples) is written as a code style analyzer (http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/AbstractConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs,11) with a code fix that supports fix all (http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/AbstractConvertAnonymousTypeToTupleCodeFixProvider.cs,23). This basically means this has all the first class analyzer/fix functionality, with the ability to be enforced on build. If we feel this should be just a code refactoring, then probably it can be rewritten that way OR we can keep the current analyzer/fix implementation, but change the diagnostic to a hidden, non-configurable diagnostic and change the fixer to explicitly not support FixAll.

Either way, I think we probably want to bring this issue to design meeting.

**Comment by CyrusNajmabadi** [comment]
https://github.com/dotnet/roslyn/pull/50536 will fix this.

**Comment by qsdfplkj** [comment]
And then why is this ok?

    Expression<Func<int, int, object>> f = (x, y) => Tuple.Create(x, y);

**Comment by CyrusNajmabadi** [comment]
> And then why is this ok?

Because that's Tuple, not ValueTuple.

**Comment by qsdfplkj** [comment]
Ok:

    Expression<Func<int, int, object>> f = (x, y) => ValueTuple.Create(x, y);

**Comment by CyrusNajmabadi** [comment]
> And then why is this ok?

Why would that not be ok?

**Comment by qsdfplkj** [comment]
The error expression tree may not contain a Rupel. But it seems only to apply to this (,) syntax. But I guess I'm missing some context. That's why a asked why is a tuole (or valuetuple) not showing this error?

**Comment by CyrusNajmabadi** [comment]
Because one is an actual tuple, the other is calls to a normal api. There's no way for an expr tree to represent the difference.
